### PR TITLE
fix: confirmation before navigating away from unsaved state (DHIS2-8432)

### PR DESCRIPTION
**Pull request by martinkrulltott**
Helps the user avoid losing unsaved work without annoying them by alerting too often. Implements confirmation checks for edited AOs, i.e. previously saved AOs that have new unsaved changes. A.k.a. 'dirty AOs'. https://jira.dhis2.org/browse/DHIS2-8432

For 'dirty AOs':
* Displays a browser `beforeunload` confirmation message when trying to go to a different URL, close the window/tab, go back to another URL etc.
* Displays a custom `ui-core` `Modal` when clicking 'Open' or 'New'

------------

Browser built-in confirmation

image

------------

Custom modal

image

### CODE REVIEW

File: packages/app/src/components/App.js [hunk 1]
```diff
@@ -152,6 +174,19 @@ export class App extends Component {
                 e.ctrlKey === true &&
                 this.props.setCurrentFromUi(this.props.ui)
         )
+
+        const t = this
```
Comment (janhenrikoverland) [inline packages/app/src/components/App.js]: Not necessary as you are using an arrow function in the event listener? You should be able to just do this.props.visualization etc inside the arrow function.

File: packages/app/src/components/App.js [hunk 2]
```diff
@@ -210,6 +245,40 @@ export class App extends Component {
                         )}
                     </div>
                 </div>
+                {this.state.locationToConfirm && (
+                    <Modal small>
+                        <ModalTitle>Discard unsaved changes?</ModalTitle>
+                        <ModalContent>
+                            Are you sure you want to leave this visualization?
```
Comment (janhenrikoverland) [inline packages/app/src/components/App.js]: i18n?

File: packages/app/src/components/App.js (hunk 1 above)
Comment (martinkrulltott) [reply packages/app/src/components/App.js]: I thought the same but it didn't work, I tried multiple combinations but this was the only way to make it work....

Comment (janhenrikoverland) [reply packages/app/src/components/App.js]: Weird, seems to work here

Comment (janhenrikoverland) [reply packages/app/src/components/App.js]: I get t === this // true

File: packages/app/src/components/App.js (hunk 2 above)
Comment (martinkrulltott) [reply packages/app/src/components/App.js]: Yes, forgot. Added now.

File: packages/app/src/components/App.js (hunk 1 above)
Comment (martinkrulltott) [reply packages/app/src/components/App.js]: Weird, also working for me now. Before I got `this` but it was different to the `this` I wanted. Maybe there was a cache issue when I tested ⚡️ anyway, updated!

**Pull request merged**